File: lib/failsafe.js
// lib/failsafe.js

const MONTHS = [
  "January","February","March","April","May","June",
  "July","August","September","October","November","December"
];

/**
 * Minimal, robust validator for CNES bin tables.
 * - Ensures at least one <table>.
 * - Optionally checks header months.
 * - Ensures one row contains the required keyword.
 *
 * @param {CheerioStatic} $ - cheerio-loaded page
 * @param {Object} options
 * @param {string[]} options.expectedMonths - month names to check, [] to skip check
 * @param {string} options.requiredKeyword - area name to validate (e.g. "Brue")
 */
export function validateBinTable($, { expectedMonths = MONTHS, requiredKeyword = "Brue" } = {}) {
  const tables = $("table");
  if (!tables.length) {
    throw new Error("No <table> found on CNES page");
  }

  // Helper: collect headers from a table
  const collectHeaders = ($table) => {
    let headers = [];
    $table.find("thead th").each((_, th) => headers.push($(th).text().trim()));
    if (headers.length === 0) {
      const firstRow = $table.find("tr").first();
      if (firstRow.length) {
        firstRow.find("th,td").each((_, cell) => headers.push($(cell).text().trim()));
      }
    }
    return headers;
  };

  // Helper: does header row look like it contains months?
  const hasExpectedMonth = (headers) => {
    if (!expectedMonths || expectedMonths.length === 0) return true;
    return headers.some((h) =>
      expectedMonths.some((m) => h.toLowerCase().includes(m.toLowerCase()))
    );
  };

  // Helper: does this table contain the keyword row?
  const tableHasKeyword = ($table, needle) => {
    let found = false;
    $table.find("tr").each((_, row) => {
      const firstCell = $(row).find("td,th").first();
      if (firstCell && firstCell.text().trim().toLowerCase().includes(needle)) {
        found = true;
        return false; // break loop
      }
    });
    return found;
  };

  const needle = String(requiredKeyword).toLowerCase();
  let headers = collectHeaders(tables.first());

  // 1) Try the first table
  let picked = null;
  if (hasExpectedMonth(headers) && tableHasKeyword(tables.first(), needle)) {
    picked = tables.first();
  } else {
    // 2) fallback scan
    tables.slice(1).each((_, table) => {
      if (picked) return;
      const h = collectHeaders($(table));
      if (hasExpectedMonth(h) && tableHasKeyword($(table), needle)) {
        picked = $(table);
        headers = h;
      }
    });
  }

  if (!picked) {
    throw new Error(`No row containing "${requiredKeyword}" found`);
  }

  if (expectedMonths.length && !hasExpectedMonth(headers)) {
    throw new Error(`No month-like header found. Saw: ${headers.join(", ")}`);
  }

  // ✅ passes silently if everything is ok
  return true;
}

File: lib/failsafe.py
# lib/failsafe.py

from bs4 import BeautifulSoup

MONTHS = [
    "January","February","March","April","May","June",
    "July","August","September","October","November","December"
]

def validate_bin_table(soup: BeautifulSoup, required_keyword="Brue", expected_months=MONTHS):
    """
    Minimal, robust validator for CNES bin tables.
    - Ensures at least one <table>.
    - Optionally checks header months.
    - Ensures one row contains the required keyword.
    """
    tables = soup.find_all("table")
    if not tables:
        raise ValueError("No <table> found on CNES page")

    def collect_headers(table):
        headers = [th.get_text(strip=True) for th in table.find_all("th")]
        if not headers:
            first_row = table.find("tr")
            if first_row:
                headers = [cell.get_text(strip=True) for cell in first_row.find_all(["th", "td"])]
        return headers

    def has_expected_month(headers):
        if not expected_months:  # skip check
            return True
        return any(
            any(m.lower() in h.lower() for m in expected_months)
            for h in headers
        )

    def table_has_keyword(table, needle):
        for row in table.find_all("tr"):
            first_cell = row.find(["td", "th"])
            if first_cell and needle.lower() in first_cell.get_text(strip=True).lower():
                return True
        return False

    needle = required_keyword.lower()
    picked = None
    headers = []

    # 1) Try first table
    headers = collect_headers(tables[0])
    if has_expected_month(headers) and table_has_keyword(tables[0], needle):
        picked = tables[0]
    else:
        # 2) fallback scan other tables
        for t in tables[1:]:
            headers = collect_headers(t)
            if has_expected_month(headers) and table_has_keyword(t, needle):
                picked = t
                break

    if not picked:
        raise ValueError(f'No row containing "{required_keyword}" found')

    if expected_months and not has_expected_month(headers):
        raise ValueError(f"No month-like header found. Saw: {headers}")

    # ✅ passes silently if everything is ok
    return True
